haskell-auto-import.el

```diff
@@ -0,0 +1,286 @@
+;;; haskell-navigate-imports.el --- A function for cycling through Haskell import lists -*- lexical-binding: t -*-
+
+;; Copyright (C) 2016  Bradley Hardy
+
+;; Author: Bradley Hardy <bradleyhardy@live.com>
+
+;; This file is not part of GNU Emacs.
+
+;; This program is free software: you can redistribute it and/or
+;; modify it under the terms of the GNU General Public License as
+;; published by the Free Software Foundation, either version 3 of
+;; the License, or (at your option) any later version.
+
+;; This program is distributed in the hope that it will be
+;; useful, but WITHOUT ANY WARRANTY; without even the implied
+;; warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+;; PURPOSE.  See the GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public
+;; License along with this program.  If not, see
+;; <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Use of the exported interactive functions should never leave the imports in
+;; an invalid state, as long as all names provided by the user exist in the
+;; right place. Adding an unqualified import twice will merge the sub-imports.
+
+;; This module works independently of most of the rest of haskell-mode, just
+;; requiring haskell-navigate-imports.
+
+;; Exports four interactive functions:
+;; 1. haskell-auto-import-unqualified
+;; 2. haskell-auto-import-unqualified-all
+;; 3. haskell-auto-import-qualified
+;; 4. haskell-auto-import-qualified-and-unqualified
+
+;; Example usage:
+
+;; (require 'haskell-auto-import)
+;; (define-key haskell-mode-map (kbd "C-c i i") 'haskell-auto-import-unqualified)
+;; (define-key haskell-mode-map (kbd "C-c i a") 'haskell-auto-import-unqualified-all)
+;; (define-key haskell-mode-map (kbd "C-c i q") 'haskell-auto-import-qualified)
+;; (define-key haskell-mode-map (kbd "C-c i x") 'haskell-auto-import-qualified-and-unqualified)
+
+;;; Code:
+
+(require 'haskell-navigate-imports)
+(require 'cl-lib)
+
+(defun haskell-auto-import-unqualified (module-name sub-imports)
+  "Add a new unqualified import of the provided module and sub-imports. If
+sub-imports is left blank, import everything in the module."
+  (interactive "MModule Name: \nMSub-Imports (comma separated, leave blank to import all): ")
+  (let* ((all-infos (haskell-auto-import-module-import-info module-name))
+         (non-qualified (member-if '(lambda (info) (not (car (cdr (cdr info))))) all-infos))
```

Comment:
Use `cl-member-if`. No need to quote lambdas. Consider using `cl-destructuring-bind` or pattern matching with `pcase` instead of nested chains of `car`s and `cdr`s.